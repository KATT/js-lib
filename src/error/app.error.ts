/**
 * Base class for all our (not system) errors
 */
export class AppError extends Error {
  constructor (message?: string, public data?: any) {
    super(message)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
