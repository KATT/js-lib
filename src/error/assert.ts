import { HttpErrorData, _deepEquals, _stringifyAny } from '..'
import { AppError } from './app.error'

/**
 * Evaluates the `condition` (casts it to Boolean).
 * Expects it to be truthy, otherwise throws AppError.
 *
 * Should be used NOT for "expected" / user-facing errors, but
 * vice-versa - for completely unexpected and 100% buggy "should never happen" cases.
 *
 * It'll result in http 500 on the server (cause that's the right code for "unexpected" errors).
 * Pass { httpStatusCode: x } at errorData argument to override the http code (will be picked up by backend-lib).
 *
 * API is similar to Node's assert(), except:
 * 1. Throws js-lib's AppError
 * 2. Has a default message, if not provided
 * 3. Sets `userFriendly` flag to true, cause it's always better to have at least SOME clue, rather than fully generic "Oops" error.
 */
export function _assert(
  condition: any, // will be evaluated as Boolean
  message?: string,
  errorData?: HttpErrorData,
): asserts condition {
  if (!condition) {
    throw new AppError(message || '_assert error (see stacktrace)', {
      userFriendly: true,
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.equals().
 *
 * Does SHALLOW, but strict equality (===), use _assertDeepEquals() for deep equality.
 */
export function _assertEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: HttpErrorData,
): asserts actual is T {
  if (actual !== expected) {
    let msg = `_assertEquals got (${actual}), but expected (${expected})`
    if (message) msg += `, ${message}`
    throw new AppError(msg, {
      userFriendly: true,
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.deepEquals().
 *
 * Does DEEP equality via _deepEquals()
 */
export function _assertDeepEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: HttpErrorData,
): asserts actual is T {
  if (!_deepEquals(actual, expected)) {
    const msg = [
      `_assertDeepEquals`,
      `got     : ${_stringifyAny(actual)}`,
      `expected: ${_stringifyAny(expected)}`,
      message,
    ]
      .filter(Boolean)
      .join('\n')

    throw new AppError(msg, {
      userFriendly: true,
      ...errorData,
    })
  }
}
