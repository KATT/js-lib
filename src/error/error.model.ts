import { StringMap } from '../types'

/**
 * Extendable payload object to transfer custom additional data with AppError.
 */
export interface ErrorData extends StringMap<any> {
  /**
   * String constant/enum that is specific for certain type of error and never changes.
   * Allows Frontend to deal with certain type of error in a specific way.
   */
  code?: string

  /**
   * User-friendly error (e.g in User's language).
   */
  userMessage?: string

  /**
   * Error id in some error tracking system (e.g Sentry).
   */
  errorId?: string
}

export interface HttpErrorData extends ErrorData {
  /**
   * @default 500
   */
  httpStatusCode: number
}

/**
 * Portable object that represents Error.
 * Has extendable generic `data` property.
 * Can be easily transformed from/to Error class via:
 * - AppError.fromErrorObject()
 * - appError.toErrorObject()
 *
 * Cannot contain any properties that stock Error doesn't have, otherwise
 * it will be hard to transform it to ErrorObject.
 * Everything "extra" should go under `data`.
 */
export interface ErrorObject<DATA_TYPE extends ErrorData = ErrorData> {
  /**
   * Name of the error / error class.
   * @default Error
   */
  name: string

  /**
   * "Technical" error message as it comes from Error.message
   */
  message?: string

  /**
   * Stacktrace of error, \n-separated, as it comes from Error.stack
   */
  stack?: string

  /**
   * Custom data to be passed with an error. Extendable.
   * It's non-optional, to save some null-checks.
   */
  data: DATA_TYPE
}

/**
 * JSON HTTP response from the Backend that represents "Error".
 */
export interface ErrorResponse<DATA_TYPE extends ErrorData = ErrorData> {
  error: ErrorObject<DATA_TYPE>
}
