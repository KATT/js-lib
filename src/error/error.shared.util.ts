import { AppError, HttpError, HttpErrorData, objectUtil } from '..'
import { ErrorObject } from '../error/error.model'

const EMPTY_STRING_MSG = 'empty_string'

class ErrorSharedUtil {
  /**
   * Converts "anything" to String error message.
   * Objects get converted to prettified JSON string.
   */
  anyToErrorMessage (o: any): string {
    if (objectUtil.isObject(o)) {
      return JSON.stringify(o, null, 2)
    }

    return String(o) || EMPTY_STRING_MSG
  }

  anyToErrorObject (o: any): ErrorObject {
    if (o instanceof Error) return this.errorToErrorObject(o)

    return {
      name: 'Error',
      message: this.anyToErrorMessage(o),
      data: { ...(o || {}).data }, // empty by default
    }
  }

  errorToErrorObject (e: Error): ErrorObject {
    if (e instanceof AppError) return this.appErrorToErrorObject(e)

    return {
      name: e.name,
      message: e.message,
      stack: e.stack,
      data: { ...(e as any).data }, // empty by default
    }
  }

  errorObjectToAppError<T> (o: ErrorObject<T>): AppError<T> {
    const err = Object.assign(new AppError(o.message, o.data), {
      // name: err.name, // cannot be assigned to a readonly property like this
      stack: o.stack,
    })

    Object.defineProperty(err, 'name', {
      value: o.name,
    })

    return err
  }

  errorObjectToHttpError (o: ErrorObject): HttpError {
    const err = Object.assign(new HttpError(o.message, o.data), {
      // name: err.name, // cannot be assigned to a readonly property like this
      stack: o.stack,
    })

    Object.defineProperty(err, 'name', {
      value: o.name,
    })

    return err
  }

  appErrorToErrorObject<T> (err: AppError<T>): ErrorObject<T> {
    return {
      name: err.name,
      message: err.message,
      stack: err.stack,
      data: err.data,
    }
  }

  appErrorToHttpError (err: AppError<HttpErrorData>): HttpError {
    return new HttpError(err.message, err.data)
  }
}

export const errorSharedUtil = new ErrorSharedUtil()
