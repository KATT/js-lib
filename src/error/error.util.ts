import { AppError, HttpError, HttpErrorData, isObject } from '..'
import { ErrorObject } from '../error/error.model'

const EMPTY_STRING_MSG = 'empty_string'

/**
 * Converts "anything" to String error message.
 * Objects get converted to prettified JSON string.
 */
export function anyToErrorMessage (o: any): string {
  if (isObject(o)) {
    return JSON.stringify(o, null, 2)
  }

  return String(o) || EMPTY_STRING_MSG
}

export function anyToErrorObject (o: any): ErrorObject {
  if (o instanceof Error) return errorToErrorObject(o)

  return {
    name: 'Error',
    message: anyToErrorMessage(o),
    data: { ...(o || {}).data }, // empty by default
  }
}

export function errorToErrorObject (e: Error): ErrorObject {
  if (e instanceof AppError) return appErrorToErrorObject(e)

  return {
    name: e.name,
    message: e.message,
    stack: e.stack,
    data: { ...(e as any).data }, // empty by default
  }
}

export function errorObjectToAppError<T> (o: ErrorObject<T>): AppError<T> {
  const err = Object.assign(new AppError(o.message, o.data), {
    // name: err.name, // cannot be assigned to a readonly property like this
    stack: o.stack,
  })

  Object.defineProperty(err, 'name', {
    value: o.name,
  })

  return err
}

export function errorObjectToHttpError (o: ErrorObject): HttpError {
  const err = Object.assign(new HttpError(o.message, o.data), {
    // name: err.name, // cannot be assigned to a readonly property like this
    stack: o.stack,
  })

  Object.defineProperty(err, 'name', {
    value: o.name,
  })

  return err
}

export function appErrorToErrorObject<T> (err: AppError<T>): ErrorObject<T> {
  return {
    name: err.name,
    message: err.message,
    stack: err.stack,
    data: err.data,
  }
}

export function appErrorToHttpError (err: AppError<HttpErrorData>): HttpError {
  return new HttpError(err.message, err.data)
}
