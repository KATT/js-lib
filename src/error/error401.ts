import { AppError } from './app.error'
import { ErrorData } from './error.model'

/**
 * HTTP 401: Unauthorized
 */
export class Error401 extends AppError {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
