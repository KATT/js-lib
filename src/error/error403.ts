import { AppError } from './app.error'
import { ErrorData } from './error.model'

/**
 * HTTP 403: Forbidden
 */
export class Error403 extends AppError {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
