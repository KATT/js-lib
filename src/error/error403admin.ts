import { ErrorData } from './error.model'
import { Error403 } from './error403'

/**
 * HTTP 403: Admin access forbidden
 */
export class Error403Admin extends Error403 {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
