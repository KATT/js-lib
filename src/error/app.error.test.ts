import { AppError } from './app.error'

test('appError properties should be present', () => {
  // Error.captureStackTrace = false as any
  const r = new AppError('hello')
  // console.log(r.message, r.name, r.stack)
  expect(r.message).toBe('hello')
  expect(r.name).toBe('AppError')
  expect(r.stack).not.toBeUndefined()

  const data = { a: 'b' }
  const r2 = new AppError('hello', data)
  expect(r2.data).toEqual(data)
})

test('appError should work when Error.captureStacktrace is n/a', () => {
  Error.captureStackTrace = false as any
  const r = new AppError('hello')
  // console.log(r.message, r.name, r.stack)
  expect(r.message).toBe('hello')
  expect(r.name).toBe('AppError')
  expect(r.stack).not.toBeUndefined()
})

test('toErrorObject / fromErrorObject snapshot', () => {
  const data = { a: 'b' }
  const err1 = new AppError('hello', data)
  const err2 = err1.toErrorObject()
  // console.log(err2)

  expect(err2).toMatchSnapshot({
    stack: expect.stringContaining('AppError'),
  })

  const err3 = AppError.fromErrorObject(err2)
  expect(err3.message).toBe('hello')
  expect(err3.name).toBe('AppError')
  expect(err3.stack).not.toBeUndefined()
  expect(err3.data).toEqual(data)
})
