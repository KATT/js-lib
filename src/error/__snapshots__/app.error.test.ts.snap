// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toErrorObject / fromErrorObject snapshot 1`] = `
Object {
  "data": Object {
    "a": "b",
  },
  "message": "hello",
  "name": "AppError",
  "stack": StringContaining "AppError",
}
`;
